ArduPlane/parachute.cpp

```diff
@@ -56,3 +59,55 @@ bool Plane::parachute_manual_release()
 
     return true;
 }
+
+// Code to detect loss of control for ArduPlane
+#define PARACHUTE_CHECK_TRIGGER_SEC         1  // 1 second of loss of control triggers the parachute
+
+/*
+  parachute_emergency_check - trigger the release of the parachute
+  automatically if a critical situation is detected
+*/
+void Plane::parachute_emergency_check()
+{
+    uint32_t now = millis();
+
+    // exit immediately if parachute or automatic release is not enabled, or already released
+    if (!parachute.enabled() || parachute.released() || g.parachute_auto_enabled == 0) {
+        chute_control_loss_ms = 0;
+        return;
+    }
+
+    // only automatically release in AUTO mode
+    if (control_mode != AUTO) {
+        chute_control_loss_ms = 0;
+        return;
+    }
+
+    // do not release if taking off or landing
+    if (auto_state.takeoff_complete == false || mission.get_current_nav_cmd().id == MAV_CMD_NAV_LAND) {
+        chute_control_loss_ms = 0;
+        return;
+    }
+
+    if (relative_altitude() < parachute.alt_min()) {
+        chute_control_loss_ms = 0;
+        return;
+    }
+
+    // do not release if we are flying within given error
+    if (altitude_error_cm < g.parachute_auto_error * 100.0f) {
+        chute_control_loss_ms = 0;
+        return;
+    }
+
+    // at this point we consider control lost
+    if (chute_control_loss_ms == 0) {
+        gcs_send_text_fmt(MAV_SEVERITY_WARNING, "Crash: Starting to lose control, %d m error", altitude_error_cm / 100);
+        chute_control_loss_ms = now;
+    } else if (now - chute_control_loss_ms > PARACHUTE_CHECK_TRIGGER_SEC * 1000) {
+        // we have lost control for too long
+        gcs_send_text_fmt(MAV_SEVERITY_WARNING, "Crash: Critical altitude error %d m", altitude_error_cm / 100);
```

Comment:
Sure, I'll change it. However, these messages are sent very very rarely and this division operation should be negligible compared to all the function calls, string formatting, and sending it to the GCS. I just thought division was more readable for me.